src/pip/_internal/network/session.py

```diff
@@ -76,136 +70,275 @@
-# For more background, see: https://github.com/pypa/pip/issues/5499
-CI_ENVIRONMENT_VARIABLES = (
-    # Azure Pipelines
-    "BUILD_BUILDID",
-    # Jenkins
-    "BUILD_ID",
-    # AppVeyor, CircleCI, Codeship, Gitlab CI, Shippable, Travis CI
-    "CI",
-    # Explicit environment variable.
-    "PIP_IS_CI",
-)
+class Telemetry:
+    """Return a string representing the user agent.
+
+    FIXME: This string is currently propagated as a header into every single HTTP
+           request pip makes.  It should really be subject to a formal PEP process
+           in order to ensure user consent around telemetry is respected.
+
+    The current implementation looks very much like the spec for *environment
+    markers*: (https://packaging.python.org/en/latest/specifications/dependency-specifiers/#environment-markers).
+
+    Environment markers were developed to provide a rich semantic structure for the
+    build environment of a dependency. Dependency resolvers like pip can then
+    manipulate these logical operators to calculate compatibility relationships,
+    *without* executing arbitrary code. This idea extends to version strings
+    themselves as well: the '===' or "arbitrary" operator was
+    specifically intended as an "escape hatch", for version strings that cannot be
+    made to conform to standard Python version strings
+    (https://packaging.python.org/en/latest/specifications/version-specifiers/#arbitrary-equality).
+
+    Environment markers can therefore be viewed as a language for codebases to
+    communicate their build process and requirements
+    (https://packaging.python.org/en/latest/flow/#the-packaging-flow) to a generic
+    resolver. Spack specs
+    (https://spack.readthedocs.io/en/latest/spec_syntax.html#sec-specs)
+    are very similar in spirit, and arose precisely to codify the recursive
+    and conditional relationships across multi-language software stacks.
+
+    Extending this concept, we could consider the *user agent id* as a language for
+    resolvers to communicate their requirements to external services and
+    execution environments. For example, clients for the GNU Make Jobserver protocol use
+    an environment variable to indicate how the jobserver should communicate to them,
+    and to indicate the maximum bandwidth they can tolerate for parallel execution:
+    (https://www.gnu.org/software/make/manual/html_node/Job-Slots.html).
+
+    The pants build tool specifically highlights the risk of exposing proprietary
+    information through thoughtless telemetry (https://www.pantsbuild.org/stable/docs/using-pants/anonymous-telemetry).
+    As a result, not only do they explicitly specify the information being recorded
+    (https://www.pantsbuild.org/stable/docs/using-pants/anonymous-telemetry#what-data-is-sent),
+    but they additionally incorporate anonymity as an explicit design goal
+    (https://www.pantsbuild.org/stable/docs/using-pants/anonymous-telemetry#how-we-ensure-anonymity).
```

Comment:
This seems all quite opinionated to be in a docstring, and doesn't recognize a malicious index could extract most of the information from a user by simply observing what wheels they download.